tests/providers/aws/e2e_cli.py

```diff
@@ -0,0 +1,332 @@
+  """
+
+  @classmethod
+  @typing.no_type_check
+  @IgnoreWarnings
+  def setUpClass(cls):
+    try:
+      project_info = utils.ReadProjectInfo(['instance', 'zone'])
+    except (OSError, RuntimeError, ValueError) as exception:
+      raise unittest.SkipTest(str(exception))
+    cls.instance_to_analyse = project_info['instance']
+    cls.zone = project_info['zone']
+    cls.dst_zone = project_info.get('destination_zone', None)
+    cls.volume_to_copy = project_info.get('volume_id', None)
+    cls.encrypted_volume_to_copy = project_info.get('encrypted_volume_id', None)
+    cls.aws = account.AWSAccount(cls.zone)
+    cls.analysis_vm_name = 'new-vm-for-analysis'
+    cls.cli = aws_cli.AWSCLIHelper()
+    cls.analysis_vm = cls._RunStartAnalysisVm(cls.analysis_vm_name, cls.zone)
+    cls.volumes = []  # List of (AWSAccount, AWSVolume) tuples
+
+  @typing.no_type_check
+  @IgnoreWarnings
+  def testBootVolumeCopy(self):
+    """End to end test on AWS.
+
+    Test copying the boot volume of an instance.
+    """
+
+    volume_copy = self._RunCreateVolumeCopy(
+        self.zone,
+        instance_id=self.instance_to_analyse
+        # volume_id=None by default, boot volume of instance will be copied
+    )
+    # The volume should be created in AWS
+    aws_volume = self.aws.ResourceApi(EC2_SERVICE).Volume(volume_copy.volume_id)
+    self.assertEqual(aws_volume.volume_id, volume_copy.volume_id)
+    self._StoreVolumeForCleanup(self.aws, aws_volume)
+
+  @typing.no_type_check
+  @IgnoreWarnings
+  def testVolumeCopy(self):
+    """End to end test on AWS.
+
+    Test copying a specific volume.
+    """
+
+    if not self.volume_to_copy:
+      return
+
+    volume_copy = self._RunCreateVolumeCopy(
+        self.zone, volume_id=self.volume_to_copy)
+    # The volume should be created in AWS
+    aws_volume = self.aws.ResourceApi(EC2_SERVICE).Volume(volume_copy.volume_id)
+    self.assertEqual(aws_volume.volume_id, volume_copy.volume_id)
+    self._StoreVolumeForCleanup(self.aws, aws_volume)
+
+  @typing.no_type_check
+  @IgnoreWarnings
+  def testVolumeCopyToOtherZone(self):
+    """End to end test on AWS.
+
+    Test copying a specific volume to a different AWS availability zone.
+    """
+
+    if not (self.volume_to_copy and self.dst_zone):
+      return
+
+    volume_copy = self._RunCreateVolumeCopy(
+        self.zone, dst_zone=self.dst_zone, volume_id=self.volume_to_copy)
+    # The volume should be created in AWS
+    aws_account = account.AWSAccount(self.dst_zone)
+    aws_volume = aws_account.ResourceApi(EC2_SERVICE).Volume(
+        volume_copy.volume_id)
+    self.assertEqual(aws_volume.volume_id, volume_copy.volume_id)
+    self._StoreVolumeForCleanup(aws_account, aws_volume)
+
+  @typing.no_type_check
+  @IgnoreWarnings
+  def testListImages(self):
+    """End to end test on AWS.
+
+    Test listing AMI images with a filter.
+    """
+
+    qfilter = 'ubuntu*'
+    images = self._RunListImages(self.zone, qfilter)
```

Comment:
Looks like _RunListImages is only called once throughout the whole file. Why make it a separate function? I think it would be more clear to have everything in the test function.